### ISSUE TITLE
Research/Implement Pokémon validation signatures

### ISSUE THREAD

**Issue by InternalLoss** [mm201/pkmn-classic-framework#116]
It seems Random Matchup fails currently after Wiimmfi fixed the matchmaking error, which @mm201 and I theorised is likely due to the signature not being valid (and confirmed the signature _is_ shared to the other peer). As such, this is likely to be a blocking issue for Random Matchup Free mode (and presumably Rating Mode? @zurgeg may now be able to check since Wiimmfi's side of the communication works without issue).

**Comment by mm201** [comment]
For reference, Jon's Gen6 server can be found here: https://github.com/PretendoNetwork/PKHaX
(they're also sending a zeroed out signature)

Apparently Gen6 is using an RSA-256 signature whose public key gets downloaded once from the server and saved. I don't *think* Gen5 downloads a key, or, at least, I've never seen any failed requests for it in my logs, but it's something we should test for. (test with a virgin savefile and see if it hits any unique endpoints)

**Comment by InternalLoss** [comment]
> I don't think Gen5 downloads a key,
I'd hope not, since there'd likely be no way to blank (and re-download) that key..

**Comment by zurgeg** [comment]
> It seems Random Matchup fails currently after Wiimmfi fixed the matchmaking error, which @mm201 and I theorised is likely due to the signature not being valid (and confirmed the signature _is_ shared to the other peer). As such, this is likely to be a blocking issue for Random Matchup Free mode (and presumably Rating Mode? @zurgeg may now be able to check since Wiimmfi's side of the communication works without issue).
> 
> Hopefully it's not an asymmetric key we don't have, or if it is we can make the DS accept our own signatures..

I'll try it in a minute.

**Comment by zurgeg** [comment]
Just tried. Looks like A) NoSSL is banned on Wiimmfi or B) I was autobanned for some reason or C) I was banned by a mod or D) melonDS is banned. Whatever it is, I can't connect to NAS with [23917](https://wiimmfi.de/error?e=23917). Which is also what happens when you're banned so...

**Comment by InternalLoss** [comment]
> melonDS is banned
you've not been banned, its probably just denied for some reason. Can you DM me your FC on Discord, and I'll ask someone?

**Comment by zurgeg** [comment]
> > melonDS is banned
> 
> you've not been banned, its probably just denied for some reason. Can you DM me your FC on Discord, and I'll ask someone?

Turns out I was using the wrong melonDS install.

**Comment by InternalLoss** [comment]
Oops ;)

Let us know how it goes!

**Comment by zurgeg** [comment]
@InternalLoss @mm201 Can confirm. PGL side is working. So yes, this does affect rating mode

**Comment by InternalLoss** [comment]
>  Can confirm. PGL side is working. So yes, this does affect rating mode
So even rating mode is affected by this? Ok, time to really figure out if it is truly the signature that's the issue..

**Comment by kuroppoi** [comment]
What we know about this:
- Signature algorithm used is `SHA1withRSA`
- Signature is over the entire encrypted party data sent to `/pokemon/validate`
- RSA public key modulus is unfortunately hardcoded and is: `D987D465E4EEAE582D017315F00EA3400C510B2E51E15D77D03ADCB25C830171F569FBD26A78DC69694DDD2CEFA4A9AAD1A0D9AA99705BF08038F57764EEA5AB7D6A3838678AEC262E952A1CDBB8E2FF68DC932E7F8E3AECD1FE5282EACA4161C2203FF098F79D6735E64414E185FBB3EC043D838D9B4B19072331C3F79857E5`

**Comment by Priestyle** [comment]
It's been a year since last comment. Any updates on this?

**Comment by scj643** [comment]
Wouldn't it be possible to use a cheat code or ROM patching to modify the public key that is used?

**Comment by Isaac0-dev** [comment]
@scj643 Yes, you can patch the rom. ~~I tried making a cheat code, but the code I patched belonged to an overlay, which traditional AR codes don't support...~~ Edit: Nevermind I came back to this and figured it out just by using a condition to check if the correct bytes were loaded.

Here's an AR code to fix random matchups. No rom patching necessary:
```
521B9A3C 1C06B5F8
021B9A3C 47702001
D0000000 00000000
```


The function that does the check is at `0x021b9a3c` in pokemon white US. 
This function returns 1 if the connection is allowed, 0 otherwise.
So, I patched the function to immediately return 1:
```asm
movs r0, #1 ; set register 0 to 1, hex: 01 20
bx lr ; return, hex: 70 47
```

I made a Python script that uses ndspy to find and patch this function. It works on any gen 5 rom (in my testing).

[[patch_bw_rom.py](https://github.com/user-attachments/files/22056925/patch_bw_rom.py)](https://gist.github.com/Isaac0-dev/88dfbe01188dea4197f8b026275d2b02)

It works!

![Image](https://github.com/user-attachments/assets/0fba15ec-cd79-41de-aaa6-b1fcb20431e2)

Edit: i recorded a video of this too: https://youtu.be/05kjAqAD91M

**Comment by Isaac0-dev** [comment]
Thought I'd expand on the signature being used here, since I learned about it myself. 
The server must send a SHA-1 hash of the player's Pokemon team, encrypted with 1024 bit RSA PKCS#1 v1.5. 
It must be encrypted using the private key that corresponds to the modulus which can be found inside the game (also can be found in previous messages from this issue).
The public exponent used for the key pair is `0x10001`.

When the server replies, the game decrypts it using the public modulus from inside the game, and compares the containing hash with it's own hash which the game calculates locally as well.
If any of those steps fail, the connection is rejected.

It's very unlikely that the private key will be found unless it's leaked or something. For now, the best method is to patch your game.

**Comment by artyfexios** [comment]
> It's very unlikely that the private key will be found unless it's leaked or something. For now, the best method is to patch your game.

I'm assuming this means that real cartridge copies of the game won't be able to do Random Battles until the key is found, is that right?

**Comment by aredspy** [comment]
> > It's very unlikely that the private key will be found unless it's leaked or something. For now, the best method is to patch your game.
> 
> I'm assuming this means that real cartridge copies of the game won't be able to do Random Battles until the key is found, is that right?


Yeah. I tried searching the svn repos that were in the Pokemon Gigaleak but I had no luck. They never committed any production private keys anywhere. Even the file has like one commit where they added the public key modulus with a benign commit message.

Funny thing is I found an email backup that talked about a CSR that had a private key attached, but the recipient noticed and asked the sender to make a new CSR with a fresh keypair without the private key for best security. 


There's still a very very very tiny chance it's in some other directory or repo, but I doubt it because AFAIK the leak didn't contain any hard source for the GTS server side code aside from some local tests with a dummy cert.

The total gigaleak size has grown to like 900Gb (lots of duplicate stuff) and I only parsed like maybe 70Gb worth (Mostly BW stuff). If I try it again, i'll probably throw an LLM at it because most of it is in Japanese with the old ISO encoding standard which was a pain to keep converting and translating when sifting through the data.

**Comment by Isaac0-dev** [comment]
Yeah that's not surprising to hear, it would have been a blunder on their part if they committed the private key.
While I do hope any further attempts to find it is successful, I think we should assume we won't find it. With that in mind, there are a few potential points of attack for tricking the game into accepting invalid signatures. At this point I think that's the most likely method to be of any use.

**Comment by artyfexios** [comment]
> Yeah that's not surprising to hear, it would have been a blunder on their part if they committed the private key. While I do hope any further attempts to find it is successful, I think we should assume we won't find it. With that in mind, there are a few potential points of attack for tricking the game into accepting invalid signatures. At this point I think that's the most likely method to be of any use.

Seeing as you managed to make random matchups work using a similar strategy, that might be the best way to go about it until the key somehow gets found. Though I do wonder how this can be implemented.

In the meantime though, would it be a good idea to add something to the website to tell others about the way you found to make random battles work? I have a feeling that others may also want to try the random matchups using the method you found but get discouraged after seeing its unsupported.

**Comment by kuroppoi** [comment]
> > Yeah that's not surprising to hear, it would have been a blunder on their part if they committed the private key. While I do hope any further attempts to find it is successful, I think we should assume we won't find it. With that in mind, there are a few potential points of attack for tricking the game into accepting invalid signatures. At this point I think that's the most likely method to be of any use.
> 
> Seeing as you managed to make random matchups work using a similar strategy, that might be the best way to go about it until the key somehow gets found. Though I do wonder how this can be implemented.
> 
> In the meantime though, would it be a good idea to add something to the website to tell others about the way you found to make random battles work? I have a feeling that others may also want to try the random matchups using the method you found but get discouraged after seeing its unsupported.

I think this has been discussed before and the general consensus was that it's not really a good solution.
It has a couple of drawbacks:
1. _It's not accessible_
AR codes & ROM patching are only available to players with the necessary means, and having to use them kinda defeats the point of the service.
2. _Queue pollution_
Even if you have patched users queueing for random matchup, you're also gonna have unpatched users attempting to use random matchup & cause failures.
3. _No legality checks_
If signature checks are entirely bypassed then cheaters will have free reign in public matches, which would degrade the experience for legitimate players.
Granted, this could be solved pretty easily by patching in a custom key instead (which has been tried--successully) but that still requires implementation.

Like @Isaac0-dev said, the ideal solution would be to either find the private key, or find an exploit that lets us replace it.
Both are probably gonna take some time, sadly.
